Index: Configure
===================================================================
RCS file: /src/local/repository/usr/local/pkgs/lsof/Configure,v
retrieving revision 1.1.1.14
retrieving revision 1.3
diff -u -p -r1.1.1.14 -r1.3
--- Configure	31 Mar 2009 15:57:46 -0000	1.1.1.14
+++ Configure	31 Mar 2009 16:27:14 -0000	1.3
@@ -4599,94 +4599,8 @@
 
       if test -r ${LSOF_INCLUDE}/sys/fs/zfs.h	# {
       then
-
-      # Check for required ZFS kernel header files.
-
-	LSOF_TMP1="dmu.h zfs_acl.h zfs_debug.h zfs_rlock.h zil.h spa.h zfs_context.h zfs_dir.h zfs_vfsops.h zio.h txg.h zfs_ctldir.h zfs_ioctl.h zfs_znode.h zio_impl.h"
-	LSOF_TMP2=""
-	for i in $LSOF_TMP1	# {
-	do
-	  if test ! -r ${LSOF_INCLUDE}/sys/$i	# {
-	  then
-	    if test "X$LSOF_TMP2" = "X"	# {
-	    then
-	      LSOF_TMP2=$i
-	    else
-	      LSOF_TMP2="$LSOF_TMP2 $i"
-	    fi	# }
-	  fi	# }
-	done	# }
-	if test "X$LSOF_TMP2" = "X"	# {
-	then
-	  LSOF_TMP4="${LSOF_INCLUDE}/sys"
-	  LSOF_TMP5=0
-	else
-	  LSOF_TMP3=`pwd`/dialects/sun/get-hdr-loc.sh
-	  if test ! -x $LSOF_TMP3	# {
-	  then
-	    echo "FATAL: can't execute: $LSOF_TMP3"
-	    rm -f $LSOF_HLP
-	    exit 1
-	  fi	# }
-	  LSOF_TMP4=`$LSOF_TMP3 $LSOF_TMP2`
-	  LSOF_TMP5=$?
-	fi	#}
-	if test $LSOF_TMP5 -eq 0	# {
-	then
-
-	# ZFS support has been requested.
-
-	  if test "X$LSOF_TMP4" = "X"	# {
-	  then
-
-	  # Use of lsof interal ZFS structure definitions has been
-	  # requested.
-
-	    LSOF_CFGF="$LSOF_CFGF -DHAS_ZFS=1"
-	  else
-
-	  # Use of ZFS header files at the returned location has
-	  # been requested.
-
-	    LSOF_CFGF="$LSOF_CFGF -DHAS_ZFS=2"
-	    if test "X$LSOF_TMP4" != "X${LSOF_INCLUDE}/sys"	# {
-	    then
-
-	    # Add the returned location unless it is ${LSOF_INCLUDE}/sys.
-
-	      LSOF_CFGF="$LSOF_CFGF -I$LSOF_TMP4"
-	    fi	# }
-	  fi	# }
-
-	  # Identify the ZFS version.
-
-	    LSOF_TMP1=0
-	    if test -x /usr/sbin/zpool	# {
-	    then
-	      LSOF_TMP2=`/usr/sbin/zpool upgrade -v|grep -i running|grep -i zfs`
-	      if test "X$LSOF_TMP2" != "X"	# {
-	      then
-	        LSOF_TMP3=`echo $LSOF_TMP2 | sed 's/^.* \([0-9][0-9]*\)\.$/\1/'`
-	        if test "X$LSOF_TMP3" != "X"	# {
-	        then
-		  LSOF_TMP1=$LSOF_TMP3
-	        fi	# }
-	      fi	# }
-	    fi	# }
-	    LSOF_CFGF="$LSOF_CFGF -DZFS_VERS=$LSOF_TMP1"
-	else
-
-	# The get-hdr-loc.sh script returned a non-zero exit value.
-	# If there was a message on STDOUT, an error was detected.
-	# if there was no message, ZFS support is not required.
-
-	  if test "X$LSOF_TMP4" != "X"	# {
-	  then
-	    echo "$LSOF_TMP3 detected the error: $LSOF_TMP4"
-	    rm -f $LSOF_HLP
-	    exit 1
-	  fi	# }
-	fi	# }
+	LSOF_CFGF="$LSOF_CFGF -DHAS_ZFS"
+	LSOF_CFGL="$LSOF_CFGL -lctf"
       fi	# }
       ;;
     *)
